nixos/lib/test-driver/src/test_driver/machine/ocr.py

```diff
@@ -0,0 +1,124 @@
+import os
+import shutil
+import subprocess
+from concurrent.futures import Future, ThreadPoolExecutor
+from pathlib import Path
+
+from test_driver.errors import MachineError
+
+
+def perform_ocr_on_screenshot(screenshot_path: Path) -> str:
+    """
+    Perform OCR on a screenshot that contains text.
+    Returns a string with all words that could be found.
+    """
+    variants = perform_ocr_variants_on_screenshot(screenshot_path, False)[0]
+    if len(variants) != 1:
+        raise MachineError(f"Received wrong number of OCR results: {len(variants)}")
+    return variants[0]
+
+
+def perform_ocr_variants_on_screenshot(
+    screenshot_path: Path, variants: bool = True
+) -> list[str]:
+    """
+    Same as perform_ocr_on_screenshot but will create variants of the images
+    that can lead to more words being detected.
+    Returns a string with words for each variant.
+    """
+    if shutil.which("tesseract") is None:
+        raise MachineError("OCR requested but `tesseract` is not available")
+
+    # Tesseract runs parallel on up to 4 cores.
+    # Docs suggest to run it with OMP_THREAD_LIMIT=1 for hundreds of parallel
+    # runs. Our average test run is somewhere inbetween.
+    # https://github.com/tesseract-ocr/tesseract/issues/3109
+    workers = max(1, int(os.process_cpu_count() / 4))
```

Comment:
As discussed in person, it might make sense to use NIX_BUILD_CORES if available.